grabbit/core.py

```diff
@@ -1,26 +1,164 @@
 from grabbit.external import six, inflect
-from grabbit.utils import natural_sort
-from os.path import join, basename, dirname, abspath, split
+from grabbit.utils import natural_sort, splitext
+from os.path import (join, basename, dirname, abspath, split, exists, islink,
+                     isabs, isdir)
 from functools import partial
+from six import string_types
 
 
 __all__ = ['File', 'Entity', 'Layout']
 
 
 class File(object):
 
-    def __init__(self, filename):
+    def __init__(self, filename, path_patterns=None):
         """
         Represents a single file.
         """
         self.path = filename
         self.filename = basename(self.path)
         self.dirname = dirname(self.path)
         self.entities = {}
+        self.path_patterns = path_patterns
+
+    def replace_entities(self, pattern):
+        new_path = pattern
+        ents = re.findall('\{(.*?)\}', pattern)
+        ents_matched = True
+        for ent in ents:
+            if ent in self.entities:
+                new_path = new_path.replace('{%s}' % ent, self.entities[ent])
+            else:
+                # An entity in the pattern is not an entity for this file
+                ents_matched = False
+
+        if ents_matched:
+            return new_path
+        else:
+            return None
+
+    def get_path(self, path_patterns=None):
+        """
+        Constructs a path for this file given this files entities and a list of
+        potential filename patterns to use.
+
+        Args:
+            path_patterns (str, list): One or more filename patterns to write
+                the file to. Entities should be represented by the name
+                surrounded by curly braces. Optional portions of the patterns
+                should be denoted by double curly braces.
+                Pattern example: 'sub-{subject}/{{var-{name}}}/{id}.csv'
+                Example result: 'sub-01/var-SES/1045.csv'
+
+        Returns:
+            A constructed path for this file based on the provided patterns.
+        """
+        if not path_patterns:
+            if self.path_patterns:
+                path_patterns = self.path_patterns
+            else:
+                return self.path
+
+        if isinstance(path_patterns, string_types):
+            path_patterns = [path_patterns]
+
+        for pattern in path_patterns:
+            new_path = pattern
+            optional_patterns = re.findall('\[(.*?)\]', pattern)
+            for optional_pattern in optional_patterns:
+                optional_chunk = self.replace_entities(optional_pattern)
+                if optional_chunk:
+                    new_path = new_path.replace('[%s]' % optional_pattern,
+                                                optional_chunk)
+                else:
+                    new_path = new_path.replace('[%s]' % optional_pattern,
+                                                '')
+
+            new_path = self.replace_entities(new_path)
```

Comment:
Haven't thought about this in detail, so I'm probably missing something, but is there a reason to call this twice, as opposed to moving the optional replacement logic inside `replace_entities` as well?